[CASSANDRA-16117] Improve docs about frozen types and invert UDT/Tuple order (#745)
apache/cassandra: doc/source/cql/types.rst

Lastly, as for :ref:`maps <maps>`, TTLs when used only apply to the newly inserted values.

.. _tuples:

Tuples
^^^^^^

> CQL also support tuples and tuple types (where the elements can be of different types). Tuple types and tuple literals

Reviewer: I would replace: `CQL also support tuples and tuple types (where the elements can be of different types).` by something like: `A Tuple is a fix-length set of values (fields) where each value can be of a different data type.`

are defined by:

.. productionlist::
   tuple_type: TUPLE '<' `cql_type` ( ',' `cql_type` )* '>'
   tuple_literal: '(' `term` ( ',' `term` )* ')'

and can be used thusly::

    CREATE TABLE durations (
        event text,
        duration tuple<int, text>,
    )

>     INSERT INTO durations (event, duration) VALUES ('ev1', (3, 'hours'));

Reviewer: `duration` is now an official Apache Cassandra data type so that example might be a bit confusing.

Unlike other "composed" types (collections and UDT), a tuple is always :ref:`frozen <frozen>` (without the need of the

> `frozen` keyword). It is not possible to update only some elements of a tuple (without updating the whole tuple).

Reviewer: I would remove `(without the need of the frozen keyword)` as I believed it changed over time and might not be true on all versions.

> `frozen` keyword) and it is not possible to update only some elements of a tuple (without updating the whole tuple).

Reviewer 2: ```suggestion
`frozen` keyword). It is not possible to update only some elements of a tuple (without updating the whole tuple).
```

> Also, a tuple literal should always have the same number of value than declared in the type it is a tuple of (some of

Reviewer 2: ```suggestion
A tuple literal must have the same number of items as its declaring type (some of
```

> those values can be null but they need to be explicitly declared as so).

Reviewer 2: ```suggestion
those values can be null but they must be explicitly declared).
```

.. _udts:

User-Defined Types
^^^^^^^^^^^^^^^^^^

CQL support the definition of user-defined types (UDT for short), which is basically a tuple on steroids - it's

> literally an extension of the class that represents a tuple. Such a type can be created, modified and removed using the

Reviewer 2: It's better to keep implementation details out of these docs, I think - I'm not sure it's necessary to specify that a UDT is a form of tuple.  Much as I like it, I think docs should also perhaps use less interesting language? (so "on steroids" might better be avoided, but just my 2c)

Author: Hey, @belliottsmith! All other comments have been addressed. Thanks a lot!
About this one... How do you feel about this?
```
CQL support the definition of user-defined types (UDT for short), which is basically an improved tuple.
```

Reviewer: The fact that a UDT inherit from a Tuple is an implementation detail that can be changed in the future. I would not mention a link between UDT and Tuple.
UDTs were introduce to help users to demormalized their data. To help them to avoid using an extra table (which is a dangerous thing in a distributed database).

Author: Hey, @blerer! Thanks for the review!
Do you have any suggestions on how should I write this? I thought the aforementioned suggestion was fine, but I'm not sure how to improve it. :confused:

Reviewer: The important part for me was to not mention the link with the `tuple` because UDTs and tuple are different in a lot of aspects.
What I had in mind was something like:
`A User Defined Type (UDT) is a set data fields where each field is named and typed. UDTs allow to store related information together within one colum. UDTs can be created, modified and removed using ...`

Author: Done, just as you suggested, @blerer! :upside_down_face:

> :token:`create_type_statement`, :token:`alter_type_statement` and :token:`drop_type_statement` described below. But

Reviewer: It is probably important to mention that a UDT is defined at the keyspace level and is not shared between keyspaces.

Author: On line 474, there's a note that states the following:
```
- A type is intrinsically bound to the keyspace in which it is created, and can only be used in that keyspace. At
  creation, if the type name is prefixed by a keyspace name, it is created in that keyspace. Otherwise, it is created in
  the current keyspace.
```

It was already there, so I didn't change a word. Do you think it's okay? :slightly_smiling_face:

Reviewer: You are right, no need to add anything. Sorry for missing that.

If the type dropped does not exist, an error will be returned unless ``IF EXISTS`` is used, in which case the operation
is a no-op.

.. _frozen:

Frozen Types
^^^^^^^^^^^^

The ``frozen`` keyword is used to change the way a collection or user-defined type column is serialized. When it is
present multiple values will be serialized as one, disabling updates on parts of UDTs or individual items of

> collections.

Reviewer: It might be good to start with some explanation about the non-frozen types to help the user understand.
Something like:

```
The ``frozen`` keyword is used to change the way a collection or user-defined type column is serialized.

`For non-frozen collections or UDTs, each value is serialized independently from the other values. This allow update or delete operations on a sub-set of the collections or UDTs values. For frozen collections or UDTs all the value are serialized as one, disabling the ability to perform partial updates on the values.`
```

To freeze a column, use the keyword, followed by the type in angle brackets, for instance::

    CREATE TABLE posts (
        id int PRIMARY KEY,
        title text,
        content text,
        tags frozen<set<text>>
    );

To insert a frozen value, it's just like a non-frozen column::

    INSERT INTO posts (id, title, content, tags)
            VALUES (1, 'Even Higher Availability with 5x Faster Streaming in Cassandra 4.0',
                    'Streaming is a process...', {'cassandra', 'availability'});

Updating a frozen column
~~~~~~~~~~~~~~~~~~~~~~~~

> As mentioned before, it's not possible to update an individual item of a collection::

Reviewer 2: A user may scroll docs rapidly, so "As mentioned before" and similar kinds of comments probably aren't helpful.

    UPDATE posts SET tags = tags - {'availability'} WHERE id = 1;

The above command would result in the following error::

    InvalidRequest: Error from server: code=2200 [Invalid query] message="Invalid operation (tags = tags -
    {'availability'}) for frozen collection column tags"

When there's a need to update, the full value must be provided::

    UPDATE posts SET tags = {'cassandra'} WHERE id = 1;

Note we're replacing the whole value, not just removing the unwanted item. The same is true for appending elements in

> a collection.

Reviewer: The `we` sound weird in a technical doc but it might be only me.
